### PR TITLE
支持 Spring Management Context 也在端口相同的情况下也共享 Tomcat

### PR DESCRIPTION
**Pull request by leewaiho**
### Motivation

Explain the context, and why you're making that change.
To make others understand what is the problem you're trying to solve.

当模块与基座在 Spring 配置文件 application.yml 中指定了相同的 management.server.port 时, 由于 managementContext 未经历 applicationContextStart 生命周期, ArkServiceInjectProcessor 未被注入到 managementContext 上下文, 导致 ArkTomcatServletWebServerFactory 中的 embeddedServerService 没有被注入. 最终在执行 getWebServer 函数时, 走向了 return super.getWebServer(initializers); . 创建了新的 Tomcat, 在后续模块的安装过程中, 又创建了绑定相同端口的 Tomcat, 出现了 端口绑定冲突异常.

### Modification

Describe the idea and modifications you've done.

- 在 ArkTomcatServletWebServerFactory 中新增了成员属性 EmbeddedServerServiceRegistry, 存储 端口->Tomcat 映射关系, 在非应用上下文中, 能根据端口复用 Tomcat
- 新增接口 EmbeddedServerServiceFactory 以及 Tomcat 默认实现 ArkTomcatEmbeddedServerServiceFactory, 用于初始化 EmbeddedServerService

### Result

Resolved or fixed #<GitHub issue number>. 

If there is no issue then describe the changes introduced by this PR.

https://github.com/sofastack/sofa-ark/issues/994

## Summary by CodeRabbit

- **New Features**
	- Introduced a flexible interface for managing embedded server services, enhancing modularity.
	- Implemented a registry for managing embedded server instances based on server ports, enabling streamlined registration and retrieval.
	- Enhanced functionality for adding and retrieving embedded server instances with port-specific management.
	- Improved server initialization logic for both Tomcat and Netty, ensuring robust dependency injection and management.
	- Added comprehensive tests for the new service management functionalities.
	- Introduced a new `RegistryServiceHolder` class to facilitate service management within tests.

- **Bug Fixes**
	- Refined context management for embedded Tomcat servers, ensuring accurate initialization and dependency handling.

- **Documentation**
	- Updated documentation to reflect new service interfaces and enhancements in embedded server management.
	- Enhanced test documentation to clarify new methods and their expected behaviors.

### CODE REVIEW

**Comment by leewaiho** [comment]
@coderabbitai summary

File: sofa-ark-plugin/web-ark-plugin/src/main/java/com/alipay/sofa/ark/web/embed/EmbeddedServerServiceRegistryImpl.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        }
+        synchronized (lock) {
+            svc = getService(port);
+            if (svc != null) {
+                return false;
+            }
+            embeddedServerServices.put(port, service);
+        }
+        return true;
+    }
```
[the remark this answers is not in the document]
Comment (leewaiho) [reply sofa-ark-plugin/web-ark-plugin/src/main/java/com/alipay/sofa/ark/web/embed/EmbeddedServerServiceRegistryImpl.java:49]: 语义不同, 不作修改

File: sofa-ark-plugin/web-ark-plugin/src/main/java/com/alipay/sofa/ark/web/embed/tomcat/ArkTomcatEmbeddedServerServiceFactory.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+public class ArkTomcatEmbeddedServerServiceFactory implements EmbeddedServerServiceFactory<Tomcat> {
+    @Override
+    public EmbeddedServerService<Tomcat> createEmbeddedServerService(Tomcat container) {
+        EmbeddedServerServiceImpl embeddedServerService = new EmbeddedServerServiceImpl();
+        if (container != null) {
+            embeddedServerService.setEmbedServer(container);
+        }
+        return embeddedServerService;
+    }
```
[the remark this answers is not in the document]
Comment (leewaiho) [reply sofa-ark-plugin/web-ark-plugin/src/main/java/com/alipay/sofa/ark/web/embed/tomcat/ArkTomcatEmbeddedServerServiceFactory.java:31]: 语义不同, 不作修改, 允许 container 为空

**Comment by leewaiho** [comment]
@coderabbitai review

File: sofa-ark-parent/support/ark-springboot-integration/ark-springboot-starter/src/main/java/com/alipay/sofa/ark/springboot/web/ArkTomcatServletWebServerFactory.java [hunk 3]
```diff
@@ -68,28 +71,65 @@
...
             return super.getWebServer(initializers);
-        } else if (embeddedServerService.getEmbedServer() == null) {
-            embeddedServerService.setEmbedServer(initEmbedTomcat());
         }
-        Tomcat embedTomcat = embeddedServerService.getEmbedServer();
+        EmbeddedServerService embedServerService;
+        if (isApplicationContext) {
+            // 仍然使用 embeddedServerService 字段存储应用上下文的 Tomcat, 避免 模块 与 基座配置文件指定的 server.port 不同
+            // 导致在使用 web-ark-plugin 的情况下使用不同的Tomcat
+            if (embeddedServerService.getEmbedServer() == null) {
```
Comment (yuanyuancin) [inline sofa-ark-parent/support/ark-springboot-integration/ark-springboot-starter/src/main/java/com/alipay/sofa/ark/springboot/web/ArkTomcatServletWebServerFactory.java:111]: 基座与模块指定不同server.port，业务直接不要依赖web-ark-plugin更好。

如果基于这个前提，可以原地扩展EmbeddedServerService.tomcat -> Map<port, Tomcat>，另外需要同步修改 netty-ark-plugin 相关逻辑

File: sofa-ark-plugin/web-ark-plugin/src/main/java/com/alipay/sofa/ark/web/embed/EmbeddedServerServiceRegistryImpl.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            return false;
+        }
+        synchronized (lock) {
+            svc = getService(port);
+            if (svc != null) {
+                return false;
+            }
+            embeddedServerServices.put(port, service);
+        }
+        return true;
```
Comment (yuanyuancin) [inline sofa-ark-plugin/web-ark-plugin/src/main/java/com/alipay/sofa/ark/web/embed/EmbeddedServerServiceRegistryImpl.java:48]: return embeddedServerServices.putIfAbsent(port, service) == null;

**Comment by yuanyuancin** [comment]
> Hi @leewaiho, welcome to SOFAStack community, Please sign [Contributor License Agreement!](http://cla.sofastack.tech)
> 
> After you signed CLA, we will automatically sync the status of this pull request in 3 minutes.

签一下CLA，不然没法merge

**Comment by leewaiho** [comment]
> > Hi @leewaiho, welcome to SOFAStack community, Please sign [Contributor License Agreement!](http://cla.sofastack.tech)
> > After you signed CLA, we will automatically sync the status of this pull request in 3 minutes.
> 
> 签一下CLA，不然没法merge

@yuanyuancin  应该已经签署了, 不知道为什么状态没同步

File: sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/web/AbstractEmbeddedServerService.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            return false;
+        }
+        return servers.putIfAbsent(port, container) == null;
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        return servers.values().iterator();
+    }
+}
```
[the remark this answers is not in the document]
Comment (leewaiho) [reply sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/web/AbstractEmbeddedServerService.java:43]: 似乎暂时没有必要

File: sofa-ark-plugin/web-ark-plugin/src/test/java/com/alipay/sofa/ark/web/embed/tomcat/EmbeddedServerServiceImplTest.java [hunk 6]
```diff
@@ -36,18 +36,18 @@ public void tearDown() {
...
-        assertEquals(tomcat, embeddedServerServiceImpl.getEmbedServer());
+        embeddedServerServiceImpl.putEmbedServer(port, tomcat);
+        assertEquals(tomcat, embeddedServerServiceImpl.getEmbedServer(port));
 
         Tomcat tomcat2 = new Tomcat();
-        embeddedServerServiceImpl.setEmbedServer(tomcat2);
+        embeddedServerServiceImpl.putEmbedServer(port, tomcat2);
         // should be still old tomcat
-        assertEquals(tomcat, embeddedServerServiceImpl.getEmbedServer());
+        assertEquals(tomcat, embeddedServerServiceImpl.getEmbedServer(port));
```
[the remark this answers is not in the document]
Comment (leewaiho) [reply sofa-ark-plugin/web-ark-plugin/src/test/java/com/alipay/sofa/ark/web/embed/tomcat/EmbeddedServerServiceImplTest.java:51]: LGTM

**Comment by leewaiho** [comment]
@coderabbitai summary

**Review by yuanyuancin — APPROVED** [review]
LGTM

**Comment by leewaiho** [comment]
@coderabbitai review

File: sofa-ark-parent/support/ark-springboot-integration/ark-springboot-starter/src/main/java/com/alipay/sofa/ark/springboot/web/ArkNettyReactiveWebServerFactory.java [hunk 7]
```diff
@@ -74,10 +75,16 @@ public ArkNettyReactiveWebServerFactory() {
 
     @Override
     public WebServer getWebServer(HttpHandler httpHandler) {
+        if (embeddedNettyService == null) {
```
[the remark this answers is not in the document]
Comment (leewaiho) [reply sofa-ark-parent/support/ark-springboot-integration/ark-springboot-starter/src/main/java/com/alipay/sofa/ark/springboot/web/ArkNettyReactiveWebServerFactory.java:78]: 因为原有的逻辑中也有这个空值判断, 我不确定是否注入后是否还会有用例会导致 embeddedServerService 的值为空, 所以在手动注入后仍保留了这个 if 判断
已补充注释说明

File: sofa-ark-parent/support/ark-springboot-integration/ark-springboot-starter/src/main/java/com/alipay/sofa/ark/springboot/web/ArkTomcatServletWebServerFactory.java [hunk 8]
```diff
@@ -68,28 +69,40 @@
...
 
-    private String                        protocol        = DEFAULT_PROTOCOL;
+    private String                protocol        = DEFAULT_PROTOCOL;
 
-    private int                           backgroundProcessorDelay;
+    private int                   backgroundProcessorDelay;
 
     @Override
     public WebServer getWebServer(ServletContextInitializer... initializers) {
+        if (embeddedServerService == null) {
```
[the remark this answers is not in the document]
Comment (leewaiho) [reply sofa-ark-parent/support/ark-springboot-integration/ark-springboot-starter/src/main/java/com/alipay/sofa/ark/springboot/web/ArkTomcatServletWebServerFactory.java:90]: 和上面一致

**Review by lvjing2 — APPROVED** [review]
LGTM

**Pull request merged**